### PR TITLE
[IR] Support "packed" mode where shape will not be padded to a power of two

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #

### CODE REVIEW

**Comment by strongoier** [comment]
\format

File: taichi/ir/snode.cpp [hunk 1]
```diff
@@ -44,13 +45,23 @@ SNode &SNode::create_node(std::vector<Index> indices,
       s = promoted_s;
     }
     TI_ASSERT(bit::is_power_of_two(s));
-    new_node.n *= s;
+    if (get_current_program().config.packed) {
```
[the remark this answers is not in the document]
Comment (strongoier) [reply taichi/ir/snode.cpp]: done

File: taichi/struct/struct_llvm.cpp [hunk 2]
```diff
@@ -192,20 +192,39 @@ void StructCompilerLLVM::generate_refine_coordinates(SNode *snode) {
   auto outp_coords = args[1];
   auto l = args[2];
 
-  for (int i = 0; i < taichi_max_num_indices; i++) {
-    auto addition = tlctx_->get_constant(0);
-    if (snode->extractors[i].num_bits) {
-      auto mask = ((1 << snode->extractors[i].num_bits) - 1);
-      addition = builder.CreateAnd(
-          builder.CreateAShr(l, snode->extractors[i].acc_offset), mask);
+  if (get_current_program().config.packed) {  // no dependence on POT
```
[the remark this answers is not in the document]
Comment (strongoier) [reply taichi/struct/struct_llvm.cpp]: done

File: taichi/transforms/utils.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#include "taichi/ir/statements.h"
+
+TLANG_NAMESPACE_BEGIN
```
[the remark this answers is not in the document]
Comment (strongoier) [reply taichi/transforms/utils.cpp]: done

File: taichi/transforms/demote_dense_struct_fors.cpp [hunk 4]
```diff
@@ -51,47 +63,69 @@ void convert_to_range_for(OffloadedStmt *offloaded) {
...
-          main_loop_var, ext.acc_offset + offset,
-          ext.acc_offset + offset + ext.num_bits);
-      start_bits[p] -= ext.num_bits;
-      auto multiplier =
-          body_header.push_back<ConstStmt>(TypedConstant(1 << start_bits[p]));
-      delta = body_header.push_back<BinaryOpStmt>(BinaryOpType::mul, delta,
-                                                  multiplier);
-      new_loop_vars[j] = body_header.push_back<BinaryOpStmt>(
-          BinaryOpType::add, new_loop_vars[j], delta);
+  if (get_current_program().config.packed) {  // no dependence on POT
```
[the remark this answers is not in the document]
Comment (strongoier) [reply taichi/transforms/demote_dense_struct_fors.cpp]: done

**Comment by strongoier** [comment]
> Awesome! Could you select a few test cases (e.g. with pointer, pure dense, etc) and test them with packed both on and off?

done

**Review by k-ye — APPROVED** [review]
LGTM!

File: taichi/ir/snode.h [hunk 5]
```diff
@@ -106,7 +114,7 @@ class SNode {
   int depth{0};
 
   std::string name;
-  int64 n{0};
+  int64 n{1};
```
[the remark this answers is not in the document]
Comment (strongoier) [reply taichi/ir/snode.h]: done

**Pull request merged**